src/lib/io/include/sol-uart.h

```diff
@@ -238,7 +251,18 @@ void sol_uart_close(struct sol_uart *uart);
  * @note Caller should guarantee that tx buffer will not be freed until
  * callback is called.
  */
-bool sol_uart_write(struct sol_uart *uart, const uint8_t *tx, unsigned int length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
+int sol_uart_write(struct sol_uart *uart, const uint8_t *tx, size_t length, void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status), const void *data);
```

Comment:
Actually the transmit callback API is more efficient IMHO: less code to write on client side, less things to maintain, and the API implementation doesn't care whether to store user data or a function pointer. 

On the receive side, IMHO the best would be to have a flag like ```new_data_available``` (containing the size of the available data), and a notification when it is **set**. Not a notification (or callback) every time there is data, which can be very frequent. The client decides when to do the read (bulk read even), for better efficiency. For instance it can start a timer and read the data once per second. If all data is read, the flag is reset.